### PR TITLE
Call Reloading Script and Jupyter Kernel API for `libyt`

### PR DESCRIPTION
**Pull request by cindytsai**
### Call Reloading Script and Jupyter Kernel API for `libyt`

- Add compile flag `--libyt_reload`, `--libyt_jupyter` in `configure.py` to enable reloading script feature and jupyter notebook feature in `libyt`.
- #194 
- [x] Where should I put _using user provided jupyter kernel connection file_ parameter at? (in `Input__Parameters` or in a separate compile option)

### CODE REVIEW

File: src/YT/YT_Inline.cpp [hunk 1]
```diff
@@ -248,12 +248,24 @@ void YT_Inline()
...
+
+#ifdef LIBYT_RELOAD
+// 5-3. activate libyt reloading script feature
+   if ( yt_run_ReloadScript("LIBYT_STOP", "RELOAD", "reload.py") != YT_SUCCESS )
+       Aux_Error( ERROR_INFO, "yt_run_ReloadScript(\"LIBYT_STOP\", \"RELOAD\", \"reload.py\") failed !!\n" );
+#endif
+
+#ifdef LIBYT_JUPYTER
+// 5-4. activate jupyter kernel
+   if ( yt_run_JupyterKernel("LIBYT_STOP", true) != YT_SUCCESS )
```
Comment (cindytsai) [inline src/YT/YT_Inline.cpp:267]: @hyschive where should I put setting `true`/`false` at? 
`true` for using user provided connection file, `false` for letting `libyt` finds empty port and connect to it automatically.

Comment (cindytsai) [reply src/YT/YT_Inline.cpp:267]: ok, done!

**Review by ChunYen-Chen — APPROVED** [review]
Great! GAMER works well with `libyt` and `libyt_jupyter`.

File: src/Main/Main.cpp [hunk 2]
```diff
@@ -211,6 +211,9 @@ char                 YT_SCRIPT[MAX_STRING];
 yt_verbose           YT_VERBOSE;
 char                 YT_FIG_BASENAME[MAX_STRING];
 int                  YT_GID_Offset[NLEVEL];
+#ifdef LIBYT_JUPYTER
+bool                 LIBYT_JUPYTER_USE_CONNECTION_FILE;
```
Comment (hyschive) [inline src/Main/Main.cpp:215]: - I suggest replacing `LIBYT` with `YT` to be consistent with other libyt variables.
- Add this runtime parameter to `example/input/Input__Parameter`
- Record this variable in HDF5.

File: src/Auxiliary/Aux_TakeNote.cpp [hunk 3]
```diff
@@ -1476,9 +1488,12 @@ void Aux_TakeNote()
...
       fprintf( Note, "Parameters of YT Inline Analysis\n" );
       fprintf( Note, "***********************************************************************************\n" );
-      fprintf( Note, "YT_SCRIPT                       %s\n",      YT_SCRIPT  );
-      fprintf( Note, "YT_VERBOSE                      %d\n",      YT_VERBOSE );
-      fprintf( Note, "YT_FIG_BASENAME                 %s\n",      YT_FIG_BASENAME );
+      fprintf( Note, "YT_SCRIPT                           %s\n",      YT_SCRIPT  );
+      fprintf( Note, "YT_VERBOSE                          %d\n",      YT_VERBOSE );
+      fprintf( Note, "YT_FIG_BASENAME                     %s\n",      YT_FIG_BASENAME );
+#     ifdef LIBYT_JUPYTER
+      fprintf( Note, "LIBYT_JUPYTER_USE_CONNECTION_FILE   %s\n",      LIBYT_JUPYTER_USE_CONNECTION_FILE ? "ON" : "OFF" );
```
Comment (hyschive) [inline src/Auxiliary/Aux_TakeNote.cpp:1495]: Replace it with `fprintf( Note, "LIBYT_JUPYTER_USE_CONNECTION_FILE   %d\n",      LIBYT_JUPYTER_USE_CONNECTION_FILE );` to match other boolean variables like `YT_VERBOSE`.

File: src/Output/Output_DumpData_Total_HDF5.cpp
Comment (hyschive) [inline src/Output/Output_DumpData_Total_HDF5.cpp:1]: Update `FormatVersion` to `2475` (you can search the keyword `2474`).

File: src/configure.py [hunk 4]
```diff
@@ -567,6 +567,18 @@ def load_arguments():
                          help="Enable the interactive mode of libyt. This activates python prompt and does not shut down a simulation when there are errors in an inline python script. Must compile libyt with INTERACTIVE_MODE. Must enable <--libyt>.\n"
                        )
 
+    parser.add_argument( "--libyt_reload", type=str2bool, metavar="BOOLEAN", gamer_name="LIBYT_RELOAD",
+                         default=False,
+                         depend={"libyt":True},
+                         help="Call the reloading script feature in libyt. This activates reloading script feature, which can reload and update the script. Must compile libyt with INTERACTIVE_MODE. Must enable <--libyt>.\n"
```
Comment (hyschive) [inline src/configure.py:573]: Simplify
"Call the reloading script feature in libyt. This activates reloading script feature, which can reload and update the script."
as
"Allow for reloading libyt scripts during runtime."

File: src/configure.py [hunk 5]
```diff
@@ -567,6 +567,18 @@ def load_arguments():
...
+                       )
+
+    parser.add_argument( "--libyt_jupyter", type=str2bool, metavar="BOOLEAN", gamer_name="LIBYT_JUPYTER",
+                         default=False,
+                         depend={"libyt":True},
+                         help="Access gamer using Jupyter Notebook / JupyterLab feature in libyt. This activates jupyter kernel during runtime, where user can access gamer using Jupyter Notebook / JupyterLab. Must compile libyt with JUPYTER_KERNEL. Must enable <--libyt>.\n"
```
Comment (hyschive) [inline src/configure.py:579]: Simplify
"Access gamer using Jupyter Notebook / JupyterLab feature in libyt. This activates jupyter kernel during runtime, where user can access gamer using Jupyter Notebook / JupyterLab."
as
"Allow for in situ analysis using Jupyter Notebook / JupyterLab through libyt."

File: src/YT/YT_Inline.cpp [hunk 6]
```diff
@@ -248,12 +248,24 @@ void YT_Inline()
    if ( yt_run_FunctionArguments( "yt_inline_inputArg", 1, "\'Dens\'" ) != YT_SUCCESS )
        Aux_Error( ERROR_INFO, "yt_run_FunctionArguments() failed !!\n" );
    if ( yt_run_Function( "yt_inline" ) != YT_SUCCESS )
-       Aux_Error( ERROR_INFO, "yt_run_Function() failed !!\n" );
+       Aux_Error( ERROR_INFO, "yt_run_Function(\"yt_inline\") failed !!\n" );
```
Comment (hyschive) [inline src/YT/YT_Inline.cpp:251]: We can also update the error message `"yt_run_FunctionArguments() failed !!\n"`

File: src/YT/YT_Inline.cpp (hunk 1 above)
Comment (hyschive) [reply src/YT/YT_Inline.cpp:267]: Should we replace `true` with `YT_JUPYTER_USE_CONNECTION_FILE`? Update the error message below as well.

**Review by hyschive — CHANGES_REQUESTED** [review]
@cindytsai Thanks for the update. I've added some minor comments.

File: src/YT/YT_Inline.cpp (hunk 1 above)
Comment (cindytsai) [reply src/YT/YT_Inline.cpp:267]: Damn, I know why, because I didn't commit... :sweat:

File: src/Auxiliary/Aux_TakeNote.cpp (hunk 3 above)
Comment (cindytsai) [reply src/Auxiliary/Aux_TakeNote.cpp:1495]: Done.

File: src/Main/Main.cpp (hunk 2 above)
Comment (cindytsai) [reply src/Main/Main.cpp:215]: Done!

**Comment by cindytsai** [comment]
@hyschive I have fixed the problem, I think it is ready for review again.

**Review by hyschive — APPROVED** [review]

**Pull request merged**